### PR TITLE
Migrate physical GC and scan_metadata to remote_storage

### PR DESCRIPTION
**Pull request by arpad-m**
Migrates most of the remaining parts of the scrubber to remote_storage:

* `pageserver_physical_gc`
* `scan_metadata` for pageservers (safekeepers were done in #8595)
* `download()` in `tenant_snapshot`. The main `tenant_snapshot` is not migrated as it uses version history to be able to work in the face of ongoing changes.
 
Part of #7547

### CODE REVIEW

File: libs/remote_storage/src/s3_bucket.rs [hunk 1]
```diff
@@ -604,6 +604,76 @@ impl RemoteStorage for S3Bucket {
...
+            }
+            Err(e) => {
+                crate::metrics::BUCKET_METRICS.req_seconds.observe_elapsed(
+                    kind,
+                    AttemptOutcome::Err,
+                    started_at,
+                );
+
+                return Err(DownloadError::Other(
+                    anyhow::Error::new(e).context("download s3 object"),
```
Comment (koivunej) [inline libs/remote_storage/src/s3_bucket.rs:660]:
```suggestion
                    anyhow::Error::new(e).context("s3 head_object"),
```

File: libs/remote_storage/src/s3_bucket.rs [hunk 2]
```diff
@@ -604,6 +604,76 @@ impl RemoteStorage for S3Bucket {
...
+                ));
+            }
+        };
+
+        let (Some(last_modified), Some(size)) = (data.last_modified, data.content_length) else {
+            return Err(DownloadError::Other(anyhow!("foobar")))?;
```
Comment (koivunej) [inline libs/remote_storage/src/s3_bucket.rs:666]: Hmm... I proposed an unexpected variant for this earlier. We wouldn't even need an anyhow for `'static str`.

File: storage_scrubber/src/lib.rs [hunk 3]
```diff
@@ -513,41 +515,36 @@ async fn list_objects_with_retries_generic(
...
-        let response_stream = match s3_client
-            .get_object()
-            .bucket(bucket_name)
-            .key(key)
-            .send()
-            .await
-        {
+    for trial in 0..MAX_RETRIES {
+        let mut buf = Vec::new();
+        let download = match remote_client.download(key, &CancellationToken::new()).await {
```
Comment (koivunej) [inline storage_scrubber/src/lib.rs:524]: could just use a single fn scoped token for shorter lines as well?

File: storage_scrubber/src/lib.rs [hunk 4]
```diff
@@ -513,41 +515,36 @@ async fn list_objects_with_retries_generic(
...
-        match response_stream
-            .body
-            .into_async_read()
-            .read_to_end(&mut body_buf)
-            .await
+        match tokio::io::copy_buf(
+            &mut tokio_util::io::StreamReader::new(download.download_stream),
+            &mut buf,
+        )
+        .await
```
Comment (koivunej) [inline storage_scrubber/src/lib.rs:538]: Isn't read_to_end still available?

File: storage_scrubber/src/checks.rs [hunk 5]
```diff
@@ -48,13 +48,13 @@ impl TimelineAnalysis {
 }
 
 pub(crate) async fn branch_cleanup_and_check_errors(
-    s3_client: &Client,
-    target: &RootTarget,
+    remote_client: &GenericRemoteStorage,
+    _target: &RootTarget,
```
Comment (koivunej) [inline storage_scrubber/src/checks.rs:52]: Hmm... Feels like it would be best to add an assertion for full s3 URLs or get rid of `_target`?

**Review by koivunej — APPROVED** [review]
I don't see any big problems.

File: libs/remote_storage/src/s3_bucket.rs (hunk 2 above)
Comment (koivunej) [reply libs/remote_storage/src/s3_bucket.rs:666]: https://github.com/neondatabase/neon/issues/7127

File: storage_scrubber/src/lib.rs (hunk 4 above)
Comment (arpad-m) [reply storage_scrubber/src/lib.rs:538]: done

File: libs/remote_storage/src/s3_bucket.rs (hunk 2 above)
Comment (arpad-m) [reply libs/remote_storage/src/s3_bucket.rs:666]: improved the error msg

File: libs/remote_storage/src/s3_bucket.rs (hunk 1 above)
Comment (arpad-m) [reply libs/remote_storage/src/s3_bucket.rs:660]: fixed

File: storage_scrubber/src/lib.rs (hunk 3 above)
Comment (arpad-m) [reply storage_scrubber/src/lib.rs]: done

**Review by koivunej — APPROVED** [review]

File: storage_scrubber/src/checks.rs (hunk 5 above)
Comment (arpad-m) [reply storage_scrubber/src/checks.rs]: got rid of target

**Pull request merged**